File: modules/whatsapp-notification-listener/ios/WhatsappNotificationListenerModule.swift
import ExpoModulesCore

public class WhatsappNotificationListenerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("WhatsappNotificationListener")
    
    Function("isPermissionGranted") { () -> Bool in
      return false
    }
    
    Function("requestPermission") { () -> Void in
      // No-op on iOS
    }
    
    Function("setAnnouncementEnabled") { (enabled: Bool) -> Void in
      // No-op
    }
    
    Function("isAnnouncementEnabled") { () -> Bool in
      return true
    }
    
    Function("speak") { (text: String) -> Void in
      // No-op
    }
  }
}

File: modules/whatsapp-notification-listener/src/WhatsappNotificationListenerModule.ts
import { NativeModule, requireNativeModule } from 'expo';

import { WhatsappNotificationListenerEvents, WhatsappVoice } from './WhatsappNotificationListener.types';

declare class WhatsappNotificationListenerModule extends NativeModule<WhatsappNotificationListenerEvents> {
  isPermissionGranted(): boolean;
  requestPermission(): void;
  setAnnouncementEnabled(enabled: boolean): void;
  isAnnouncementEnabled(): boolean;
  speak(text: string): void;
  getAvailableVoices(): Promise<WhatsappVoice[]>;
  setSelectedVoice(voiceName: string): void;
  getSelectedVoice(): string;
}

export default requireNativeModule<WhatsappNotificationListenerModule>('WhatsappNotificationListener');
